Mathlib/Computability/LeftmostDerivation.lean

```diff
@@ -0,0 +1,409 @@
+/-
+Copyright (c) 2024 Tobias Leichtfried. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Tobias Leichtfried
+-/
+import Mathlib.Computability.ContextFreeGrammar
+
+/-!
+# Leftmost Deriviations in Context-Free Gammars
+
+This file contains the definition of a leftmost derivation. These are derivations
+where in each rewriting step the leftmost nonterminal instead of an arbitrary one is replaced
+
+## Main Definitions
+* `ContextFreeRule.RewritesLeftmost`: Leftmost counterpart to `ContextFreeRule.Rewrites`
+* `ContextFreeGrammar.ProducesLeftmost`: Leftmost counterpart to `ContextFreeRule.Produces`
+* `ContextFreeGrammar.DerivesLeftmost`: Leftmost counterpart to `ContextFreeRule.Derives`
+
+## Main Result
+* `ContextFreeGrammar.derives_leftmost_iff` : A string of terminals can be derived from a string of
+  symbols iff it can be leftmost derived.
+-/
+
+universe uT uN
+variable {T : Type uT}
+variable {N : Type uN}
+
+namespace ContextFreeRule
+open Symbol
+
+/-- Inductive definition of a single application of a given context-free rule `r` to a string `u`;
+`r.RewritesLeftmost u v` means that the `r` sends `u` to `v` by replacing the leftmost nonterminal
+symbol -/
+inductive RewritesLeftmost (r : ContextFreeRule T N) : List (Symbol T N) → List (Symbol T N) → Prop
+  /-- The replaced nonterminal is the leftmost symbol -/
+  | head (s : List (Symbol T N)) :
+      r.RewritesLeftmost (Symbol.nonterminal r.input :: s) (r.output ++ s)
+  /-- There are terminals further left than the replaced symbol -/
+  | cons (x : T) {s₁ s₂ : List (Symbol T N)} (hrs : RewritesLeftmost r s₁ s₂) :
+      r.RewritesLeftmost (terminal x :: s₁) (terminal x :: s₂)
+
+theorem RewritesLeftmost.exists_parts {r : ContextFreeRule T N}
+    {u v : List (Symbol T N)} (hr : r.RewritesLeftmost u v) :
+      ∃ (p : List T) (q : List (Symbol T N)),
+        u = p.map terminal ++ [nonterminal r.input] ++ q ∧
+        v = p.map terminal ++ r.output ++ q := by
+  induction hr with
+  | head s =>
+    use [], s
+    simp
+  | cons x _ ih =>
+    obtain ⟨p, q, hpq⟩ := ih
+    use x::p, q
+    simp [hpq]
+
+theorem RewritesLeftmost.rewrites_leftmost_of_exists_parts (r : ContextFreeRule T N)
```

Comment:
In all theorem names, `rewrites_leftmost` should be written together because `rewritesLeftmost` clearly indicates that `RewritesLeftmost` is the definition it refers to.